lib: modify NativeModule to use compileFunction

ryzokuken opened the pull request:
Modify the NativeModule class to use compileFunction instead of manually
wrapping the string source code and then compiling it.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

/cc @devsnek @bcoe

ryzokuken committed:
lib: modify NativeModule to use compileFunction

Modify the NativeModule class to use compileFunction instead of manually
wrapping the string source code and then compiling it.

bcoe commented:
@ryzokuken awesome, thank you for doing this (I'm excited to see if it fixes some of the issues I was seeing around instrumenting internal modules for coverage).

I will get some feedback to you tonight.

ryzokuken commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +258,29 @@
-      const script = new ContextifyScript(
-        source, this.filename, 0, 0,
-        cache, false, undefined
+      const fn = internalBinding('contextify').compileFunction(
```
Umm, it's only called once? Had it been called multiple times, I would have refactored it into a single destructure statement up top.

devsnek replied:
its called every time we load an internal module, and we have a great many of those.

ryzokuken replied:
@devsnek Oh wait, I get it now, sorry. Will change this.

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

bcoe reviewed (approved):
this seems to do the trick, and remove the necessity to take intro account wrapper length when calculating coverage ([confirmed here](https://github.com/bcoe/node-coverage-debug)).

It doesn't (as I'd hoped it might) solve the issue described here https://github.com/nodejs/node/issues/22919#issuecomment-432468192

jdalton commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,29 @@
+        cache,
+        false,
+        undefined,
+        [],
```
☝️ I think `contextExtensions` can be `undefined` too.

ryzokuken replied:
@jdalton I used an empty array since that's the default value I used in `vm.compileFunction` and wanted to make this as consistent as I could. I just rechecked the code to make sure, the two *should* be functionally similar, resulting in an empty `std::vector<Local<Object>> context_extensions`.

jdalton replied:
My point was more reducing throw away object creation.

ryzokuken committed:
vm: report cache rejection in vm.compileFunction

Add functionality in vm.compileFunction/CompileFunction to report if
cache data was provided yet rejected by v8.

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,29 @@
+        false,
+        undefined,
+        [],
+        ['exports', 'require', 'module', 'process', 'internalBinding']
```
I'd prefer we use a variable to refer to the function arguments somehow before passing them into the `compileFunction`, for better readability (I like that the the previous `NativeModule.wrapper` is self-explanatory)...or put a comment for each one for these arguments?

jdalton replied:
:+1: Yep, I had to lookup the args again when checking it 🙃

joyeecheung replied:
Missing an comment about the last argument here?

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -120,7 +120,6 @@
     this.exportKeys = undefined;
     this.loaded = false;
     this.loading = false;
-    this.script = null;  // The ContextifyScript of the module
```
This breaks the code cache generation (although it is already broken by https://github.com/nodejs/node/pull/23025). An easy way to test it is to run `./node --expose-internals tools/generate_code_cache.js  /tmp/node_code_cache.cc`. To fix it we need to have another way to expose the wrapped source again somehow and compile their cache in `internal/bootstrap/cache.js`.

(I'll fix the script and add a test)

bcoe replied:
should we perhaps add a `// TODO(@joyeecheung): figure out way to expose wrapped source for code cache`; or do you see this as a blocker for @ryzokuken's work.

ryzokuken replied:
For the record, I'm okay either way, I could look into it in another PR, leave it for @joyeecheung to do or add it in this one.

I'd not prefer to block this for it though since it's also kinda broken?

joyeecheung replied:
I didn’t mean to block this so this is just a comment :) I am good with leaving a TODO here

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

jdalton commented on src/node_contextify.cc:
```diff
@@ -1099,12 +1099,23 @@ void ContextifyContext::CompileFunction(
+                 Boolean::New(isolate, source.GetCachedData()->rejected))
+            .IsNothing())
+      return;
+  }
```
@ryzokuken 👆 That's a neat way to handle this. What do you think about having `compileFunction` return a results object with a `value` property (kind of like the generator protocol) but with other properties like the `cachedDataRejected` hanging off it (so it doesn't have to hang off the generated function).

ryzokuken replied:
@jdalton that's actually what I had originally intended to do when I was designing `vm.compileFunction`, but chose to do things this way for the sake of consistency (all related functions in the vm module seem to do things this way?).

I'm still not opposed to the idea of doing things that way, if enough people agree, but let's just keep in mind that that'll have to happen as a semver-major change.

Do you want me to make an issue to track people's opinions on the idea and follow it up with a PR if everyone agrees?

jdalton replied:
@ryzokuken 
> Do you want me to make an issue to track people's opinions on the idea and follow it up with a PR if everyone agrees?

That would be great, thank you!

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,29 @@
+        cache,          // cached data
+        false,          // produce cached data
+        undefined,      // parsing context
+        undefined,      // context extensions
```
Missing a comment about the last argument here?

ryzokuken replied:
I thought it was pretty self-explanatory plus was long, will add one just to be sure.

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,30 @@
       const cache = codeCacheHash[id] &&
         (codeCacheHash[id] === sourceHash[id]) ? codeCache[id] : undefined;
 
```
```suggestion
      const params = ['exports', 'require', 'module', 'process', 'internalBinding'];
```

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,30 @@
+        false,          // produce cached data
+        undefined,      // parsing context
+        undefined,      // context extensions
+        // arguments to the function
```
```suggestion
        params // formal parameters of the function
```

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,30 @@
+        undefined,      // parsing context
+        undefined,      // context extensions
+        // arguments to the function
+        ['exports', 'require', 'module', 'process', 'internalBinding']
```
```suggestion
```

joyeecheung reviewed (commented):
Made a suggestion about commenting on the params. Let's try GitHub beta magic!

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

ryzokuken commented:
@joyeecheung done! PTAL.

joyeecheung commented on lib/internal/bootstrap/loaders.js:
```diff
@@ -270,29 +259,32 @@
-      const script = new ContextifyScript(
-        source, this.filename, 0, 0,
-        cache, false, undefined
+      const params = ['exports', 'require', 'module', 'process', 'internalBinding'];
```
Now come to think of it, to reproduce the cache in the user land, we'll need to expose the params as well (otherwise we'll need to copy-paste)..Can you put this onto `NativeModule.parameters` (or something with a better name?)

devsnek replied:
make sure you Object.freeze that

ryzokuken replied:
@devsnek right. `ETOOMUCHMONKEYPATCHING`.

targos replied:
Wait, `NativeModule` is accessible to userland?

ryzokuken replied:
I guess it's not, now that I think about it. But still, better be safe than be sorry, I guess. 🤷‍♂️

ryzokuken committed:
fixup! lib: modify NativeModule to use compileFunction

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/18227/

joyeecheung commented on src/node_contextify.cc:
```diff
@@ -1099,12 +1099,23 @@ void ContextifyContext::CompileFunction(
           buf.ToLocalChecked()).IsNothing()) return;
     }
+
+    // Report if cache was produced.
```
I looked into the implementation a bit, and I wonder if we can just wrap `ScriptCompiler::CreateCodeCacheForFunction` in another binding that takes a function as argument? That way we can just manipulate the return values in the JS land, and also enable the code cache generator to use the exact function compiled to generate the cache if we simply return `fn` in `NativeModule.prototype.compile` - also, we don't actually need that argument handling complexity in the `compileFunction` C++ binding, `vm.compileFunction` can handle `produce_cached_data` in JS land and `NativeModule.prototype.compile` doesn't even need it at all.

joyeecheung replied:
Also I just noticed that the docs of `vm.compileFunction` did not mention its return value. Is that intentional?

joyeecheung replied:
(and the returned values are not tested either?)

Trott commented:
@ryzokuken Is this still being worked on? (If yes, should we add an `in progress` label? If not, should we close it?)

ryzokuken commented:
@Trott it is! The discussion was moved into another sub-change which should be merged soon and unblock this one.

bcoe commented:
@joyeecheung @ryzokuken I believe internal modules no longer have a wrapper prefix? is this work still needed?

Seems like the important issue is https://github.com/nodejs/node/pull/21573

ryzokuken commented:
@bcoe IIRC, @joyeecheung took care of this. Working on merging the other one in.

BridgeAR commented:
Ping @joyeecheung (see question above)

BridgeAR commented:
Wrong PR

Trott commented:
This looks like it may have been abandoned. Should we close this? We can re-open it if someone starts to work on it again?

ryzokuken commented:
@Trott I had a talk with @joyeecheung about these, and looks like they still need some work. Please leave it open and I'd try to look deeper into it either later today or next week.

gireeshpunathil commented:
ping @ryzokuken to see if it is moving

ryzokuken commented:
I'm super sorry, but I've had no time as of late to finish this. If you don't mind leaving this open, I can try to look into it during vacations later this month or someone else would be interested to pick it up?

bnb commented:
@ryzokuken any chance you're able to get to this?

ryzokuken commented:
@bnb sorry about that, I'll try to get back to it over the weekend.

aduh95 commented:
This needs a rebase.
